File: 백준/10799 쇠막대기.py
import sys
input = sys.stdin.readline

batch = input().strip()
stack = []
answer = 0

for i in range(len(batch)):
    if batch[i] == "(":
        stack.append(batch[i])
    else:
        if batch[i - 1] == "(":
            stack.pop()
            answer += len(stack)
        else:
            stack.pop()
            answer += 1
print(answer)

File: 백준/10799 쇠막대기.cpp
#include <bits/stdc++.h>

using namespace std;

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);

    stack<pair<int, char>> stack;
    string batch;

    cin >> batch;

    int answer = 0, count = 0;
    int index;
    
    for (int i = 0; i < batch.size(); i++) {
        if (batch[i] == '(') {
            stack.push(make_pair(i, batch[i]));
            count += 1;
        } else if (stack.top().second == '(' && batch[i] == ')') {
            index = stack.top().first;
            stack.pop();
            if (index + 1 == i) {
                count -= 1;
                answer += count;
            } else {
                answer += 1;
                count -= 1;
            }
        }
    }

    cout << answer;

}